KAFKA-19297: Refactor AsyncKafkaConsumer's use of Java Streams APIs in critical sections

kirktrue opened the pull request:
Profiling has shown that using the Collections Streams API approach adds
unnecessary overhead compared to a traditional for loop. Minor revisions
to the code have been made to use simpler constructs to improve
performance.

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Andrew Schofield
 <aschofield@confluent.io>

kirktrue committed:
KAFKA-19297: Refactor AsyncKafkaConsumer's use of Java Streams APIs in critical sections

kirktrue committed:
Reverting some overzealous changes

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java:
```diff
@@ -1135,7 +1134,9 @@ CompletableFuture<Void> revokePartitions(Set<TopicPartition> partitionsToRevoke)
-        log.info("Revoking previously assigned partitions {}", revokedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+
+        if (log.isInfoEnabled())
+            log.info("Revoking previously assigned partitions {}", Utils.sortedTopicPartitions(revokedPartitions));
```
This is actually sorting the list of partitions whereas it previously was not. That seems accidental in a PR that is trying to reduce CPU usage.

lianetm replied:
agree, and btw, this should be already sorted I expect (all these partition sets are internally `SortedSet` exactly for this, better logs).  We can probably change the input param type to `SortedSet<TopicPartition>` in this func

kirktrue replied:
I will revert the change to make everything sorted, if it wasn't already.

kirktrue replied:
Reverted change to make everything sorted.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java:
```diff
@@ -20,12 +20,41 @@
+    }
+
+    private static String join(SortedSet<?> elements) {
+        StringBuilder b = new StringBuilder();
```
If you constructed an implementation of `Iterator` which executed `Object.toString()` on its elements, you could then use `String.join` directly. I wonder what you think about using the Java library to do the string concatenation.

kirktrue replied:
I made a change to use an `Iterable`/`Iterator` under the covers to leverage `String.join()`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java:
```diff
@@ -84,16 +84,21 @@ public TopicMetadataRequestManager(final LogContext context, final Time time, fi
+                expiredRequests.add(requestState);
+        }
+
         expiredRequests.forEach(TopicMetadataRequestState::expire);
```
Couldn't you just not build `expiredRequests` and call `requestState.expire()` at L91?

kirktrue replied:
Because the request’s `expire()` method removes itself from `inflightRequests`, won't that throw an exception because the for loop's underlying `Iterator` wasn't used to remove the entry?

kirktrue replied:
Well, I take it back. Apparently it's not an error to remove it directly from the collection while iterating over the collection 🤔

kirktrue replied:
I've removed the superfluous list buildup.

lianetm replied:
> Apparently it's not an error to remove it directly from the collection while iterating over the collection

Agree it doesn't seem to fail with the tests we have, but still I would expect it may not be safe to modify the collection from within the loop (and may lead to ConcurrentModification exception at some point?). 

I would say we play safe and use an iterator, makes sense?

kirktrue replied:
Done.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. A few comments to address.

kirktrue committed:
No longer sorting topic partitions and coercing objects in order to use String.join()

kirktrue committed:
Removing unnecessary list for expiration; just expiring directly

kirktrue commented:
@AndrewJSchofield @lianetm—Anything else I need to fix before we can merge? Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java:
```diff
@@ -470,7 +475,19 @@ protected Map<Node, FetchSessionHandler.FetchRequestData> prepareFetchRequests()
+        Map<Node, FetchSessionHandler.FetchRequestData> map = new HashMap<>(fetchable.size());
+
+        for (Map.Entry<Node, FetchSessionHandler.Builder> entry : fetchable.entrySet()) {
+            Node node = entry.getKey();
+            FetchSessionHandler.FetchRequestData fetchRequestData = entry.getValue().build();
+            map.put(node, fetchRequestData);
```
```suggestion
            map.put(entry.getKey(), entry.getValue().build());
```

kirktrue replied:
This was just to make the code a little clearer about the exact processing and types in use. I'll make it more succinct.

kirktrue replied:
Changed as requested.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManager.java:
```diff
@@ -414,8 +413,8 @@ public CompletableFuture<Void> signalPartitionsRevoked(Set<TopicPartition> parti
-        if (!revokePausedPartitions.isEmpty()) {
-            log.info("The pause flag in partitions [{}] will be removed due to revocation.", revokePausedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+        if (!revokePausedPartitions.isEmpty() && log.isInfoEnabled()) {
+            log.info("The pause flag in partitions [{}] will be removed due to revocation.", Utils.topicPartitionString(revokePausedPartitions));
```
Given that er end up printing the string in between [], can't we simply print the set directly?
```suggestion
            log.info("The pause flag in partitions {} will be removed due to revocation.", revokePausedPartitions);
```

kirktrue replied:
You're right. Changed.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java:
```diff
@@ -1135,7 +1134,9 @@ CompletableFuture<Void> revokePartitions(Set<TopicPartition> partitionsToRevoke)
-        log.info("Revoking previously assigned partitions {}", revokedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+
+        if (log.isInfoEnabled())
+            log.info("Revoking previously assigned partitions {}", Utils.topicPartitionString(revokedPartitions));
```
Here also I expect we can simplify and just print the set? Only difference is that it will include the brackets, which is ok I expect?

kirktrue replied:
Yep. Changed.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:
```diff
@@ -362,12 +368,18 @@ void cleanup() {
+            CompletableApplicationEvent<?> cae = (CompletableApplicationEvent<?>) ce;
+
+            if (cae.requireSubscriptionMetadata())
+                subscriptionMetadataEvent.add(cae);
```
maybe simplify ?
```suggestion
            if (ce instanceof CompletableApplicationEvent && ((CompletableApplicationEvent<?>) ce).requireSubscriptionMetadata()) {
                subscriptionMetadataEvent.add((CompletableApplicationEvent<?>) ce);
            }
```

kirktrue replied:
Done.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java:
```diff
@@ -77,11 +81,14 @@ public Exception invokePartitionsAssigned(final SortedSet<TopicPartition> assign
-                log.info("The pause flag in partitions [{}] will be removed due to revocation.", revokePausedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+
+        if (!revokePausedPartitions.isEmpty() && log.isInfoEnabled())
+            log.info("The pause flag in partitions [{}] will be removed due to revocation.", Utils.topicPartitionString(revokePausedPartitions));
```
similar to above, if we are adding [] to the string of partitions, couldn't we just print the set?

kirktrue replied:
Done.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java:
```diff
@@ -103,11 +114,14 @@ public Exception invokePartitionsRevoked(final SortedSet<TopicPartition> revoked
-            log.info("The pause flag in partitions [{}] will be removed due to partition lost.", lostPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+
+        if (!lostPausedPartitions.isEmpty() && log.isInfoEnabled())
+            log.info("The pause flag in partitions [{}] will be removed due to partition lost.", Utils.topicPartitionString(lostPartitions));
```
ditto

kirktrue replied:
Done

kirktrue committed:
Merge branch 'trunk' into KAFKA-19297-refactor-away-streams

kirktrue committed:
Merge branch 'trunk' into KAFKA-19297-refactor-away-streams

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java:
```diff
@@ -310,7 +311,9 @@ private void process(final AssignmentChangeEvent event) {
 
-        log.info("Assigned to partition(s): {}", event.partitions().stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+        if (log.isInfoEnabled())
+            log.info("Assigned to partition(s): {}", Utils.topicPartitionString(event.partitions()));
```
ok with the change, even though this one is not on any hot-path I expect, right? (just API calls to consumer.assign that do a single trip to the background, no requests).

kirktrue replied:
True, it's just cleaner now.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableEventReaper.java:
```diff
@@ -166,9 +161,48 @@ public boolean contains(CompletableEvent<?> event) {
-                .collect(Collectors.toList());
+        List<CompletableEvent<?>> events = new ArrayList<>();
+
+        for (CompletableEvent<?> event : tracked) {
```
Should we add a comment saying that we're intentionally using for loop over the Streams API for perf, as this runs on every iteration of the background thread? (I'm afraid folks will be tempted to change this simple loops back to streams api)

kirktrue replied:
Agreed. I added one here and similar comments in a few other places.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableEventReaper.java:
```diff
@@ -166,9 +161,48 @@ public boolean contains(CompletableEvent<?> event) {
+
+            if (event.future().completeExceptionally(error)) {
+                log.debug("Event {} completed exceptionally since the consumer is closing", event);
+                count++;
```
incrementing here is not exactly the same that it used to happen before, right? 

Before the PR, the count inc would happen after the `peek(expireEvent)`, so it would count events that were expired but not completedExceptionally (the ones going through line 202, where now we don't increment the count)

This count doesn't seem used anyways, but better to be accurate and consistent with what it's done in the `reap` func (there the count includes all expired events, no matter if we completed them exceptionally or not) https://github.com/apache/kafka/blob/2b2552da1fbcdb687c99a03a81034c8f78e86871/clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableEventReaper.java#L115

kirktrue replied:
Yes, it's different. I assert it's more accurate, but I changed it to mimic the existing counting.

lianetm reviewed (commented):
Thanks @kirktrue!

kirktrue committed:
Merge branch 'trunk' into KAFKA-19297-refactor-away-streams

kirktrue committed:
PR review cleanup

kirktrue committed:
Reverted some whitespace

kirktrue committed:
Whitespace change

kirktrue committed:
Added comments and reverted more whitespace changes

lianetm reviewed (commented):
Thanks @kirktrue! Took another pass, just one comment

kirktrue committed:
Update TopicMetadataRequestManager.poll() to remove the request state via an iterator

This is to prevent possible state issues that could arise from removing elements of the list as it's being iterated over.

lianetm reviewed (approved):
Thanks! LGTM.
